### PR TITLE
curl: support VLAN Priority: --vlan-priority

### PR DESCRIPTION
**Pull request by orgads**
Add --vlan-priority option to the command line tool for setting VLAN priority.

### CODE REVIEW

**Comment by orgads** [comment]
What's the `hyper` check? It failed and there are no logs whatsoever.

**Comment by icing** [comment]
> What's the `hyper` check? It failed and there are no logs whatsoever.

Looks like github had a problem. When the CI workflow is done, try restarting the failed jobs.

File: docs/cmdline-opts/vlan-priority.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+Multi: single
+See-also:
+  - ip-tos
+Example:
+  - --vlan-priority 4 $URL
+---
+
+# `--vlan-priority`
+
+Set VLAN priority as defined in IEEE 802.1Q. (Added in 8.9.0).
```
Comment (bagder) [inline docs/cmdline-opts/vlan-priority.md:19]: Maybe mention that this sets a field on Ethernet level and maybe something about the limitations that brings? Ie that this only really works within a local network.

File: src/tool_operate.c [hunk 2]
```diff
@@ -195,6 +196,18 @@ static int sockopt_callback(void *clientp, curl_socket_t curlfd,
...
     }
   }
+#endif
+#ifdef SO_PRIORITY
+  if(config->vlan_priority > 0) {
+    int priority = (int)config->vlan_priority;
+    if(setsockopt(curlfd, SOL_SOCKET, SO_PRIORITY,
+      (const char *)&priority, sizeof(priority)) != 0) {
+      int error = errno;
+      warnf(config->global, "SO_PRIORITY %d failed with errno %d: %s;\n",
```
Comment (bagder) [inline src/tool_operate.c:206]: The user will not know what `SO_PRIORITY` is. They used `--vlan-priority`.

Comment (orgads) [reply src/tool_operate.c:206]: Reworded.

File: docs/cmdline-opts/vlan-priority.md (hunk 1 above)
Comment (orgads) [reply docs/cmdline-opts/vlan-priority.md:19]: Done

**Comment by orgads** [comment]
These flaky tests are really frustrating...

**Review by bagder — APPROVED** [review]

File: src/tool_operate.c [hunk 3]
```diff
@@ -2253,13 +2266,19 @@ static CURLcode single_transfer(struct GlobalConfig *global,
...
 #else
-          warnf(config->global,
-                "Type of service is not supported in this build.");
+          if(config->ip_tos > 0) {
+            warnf(config->global,
+                  "Type of service is not supported in this build.");
+          }
+          if(config->vlan_priority > 0) {
+            warnf(config->global,
+                  "VLAN priority is is not supported in this build.");
```
Comment (michaelgruenewald) [inline src/tool_operate.c:2280]: typo:
```suggestion
                  "VLAN priority is not supported in this build.");
```

Comment (orgads) [reply src/tool_operate.c]: Nice catch! Fixed.

**Comment by orgads** [comment]
I moved the "unsupported" handling to the callback (in a separate commit). If you like it, you can leave it or squash. Otherwise, I'll discard this commit (probably tomorrow night).

**Comment by bagder** [comment]
> I moved the "unsupported" handling to the callback (in a separate commit). If you like it, you can leave it or squash. Otherwise, I'll discard this commit (probably tomorrow night).

I prefer to detect the problem in the earlier stage rather than delaying and erroring out after having done more processing, when it is possible. For that reason prefer the previous take a little more.

Also, I think we should use `errorf()` instead of `warnf()` for this since we bail out and return an error for this case.

**Comment by orgads** [comment]
> > I moved the "unsupported" handling to the callback (in a separate commit). If you like it, you can leave it or squash. Otherwise, I'll discard this commit (probably tomorrow night).
> 
> I prefer to detect the problem in the earlier stage rather than delaying and erroring out after having done more processing, when it is possible. For that reason prefer the previous take a little more.
> 
> Also, I think we should use `errorf()` instead of `warnf()` for this since we bail out and return an error for this case.

Done. Better now?

**Review by bagder — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Comment by Cleanfastllc** [comment]
Cool

**Pull request closed without merging**